[el] add form_of parsing for inflections (nouns, adjectives)

daxida opened the pull request:
Closes #1122

- [x] Adds `form_of` and `raw_tags` for `[gender of] + πτώσ(η|εις)` templates. 
This should cover most nouns and adjectives.
- [x] Test it.
- [x] Support verbs.
- [x] Localize to English and populate `tags` instead of `raw_tags`

Notes:
- Glosses are unchanged based on this [structure]
(https://github.com/tatuylonen/wiktextract/issues/1122#issuecomment-2803472770).
- Glosses are not tested in order to avoid expanding templates. I did verify locally that indeed they were unaffected by the changes.
- Some `raw_tags` (cf. θηλυκό) are parsed from `raw_tags` into `tags` at some other points of the code. I didn't add any extra logic myself.
- I didn't see how to apply @xxyzz [comment](https://github.com/tatuylonen/wiktextract/issues/1122#issuecomment-2800936424). Everything is done at `parse_gloss_inflections`.

Let me know what you think.

kristian-clausal commented:
I will be on vacation for two weeks starting next week, so I will not be looking at these then, and we've got some other things going on during this shorter week so things will have to wait until then. Please don't do too much work during that time, I suspect much of it is not going to be appropriate / idiomatic and would need to be changed a lot. Checking and working these PR requests would be easier to do it in small portions where I can massage it together (if possible) with what I'm currently working at, which overlaps with this.

The most useful bits from this (at a quick glance) is the translation of how template names are structured. If you can give us data on that and other things, like tag translations, that would be useful. I'll comment on this, but I'm not promising to integrate it.

kristian-clausal commented on tests/test_el_inflection.py:
```diff
@@ -0,0 +1,119 @@
+            self.assertIn("glosses", sense)
+            sense.pop("glosses")
+
+            # The order of tags may change between runs
```
The order should be deterministic, yeah. For `raw_tags`, order of processing is fine. `tags` should be `.sorted()`

kristian-clausal commented on tests/test_el_inflection.py:
```diff
@@ -0,0 +1,119 @@
+        raw = """* {{θηλ_του-πτώσειςΟΑΚεν|μικρός}}"""
+        expected = [
+            {
+                "tags": ["θηλυκό"],  # This is parsed somewhere else I guess
```
Should not be in `tags`, `tags` should only be valid, English tag-data. Might be my fault.

daxida replied:
I recon this is weird but it's not something I implemented. I just add the `θηλυκό` (feminine) tag to `raw_tags` and I imagine that somewhere else it is promoted to tag. 

I had to do this to pass the test, but for instance the `ουδέτερο` (neuter) tag that appears in another test does not receive the same treatment.

kristian-clausal commented on tests/test_el_inflection.py:
```diff
@@ -0,0 +1,119 @@
+    def mktest_sense(self, raw: str, expected: list[dict[str, Any]]) -> None:
+        """Compare only senses["tags"] and senses["form_of"].
+
+        Do not compare senses["glosses"] but check that it exists. This
```
You can also do mock template data. There are examples in other tests, they're a bit finicky... Or, you can accept the broken output, as long as the tests are consistent, if it's not important for the test.

kristian-clausal commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -522,15 +515,38 @@ def parse_gloss_inflections(
+    # https://el.wiktionary.org/wiki/Πρότυπο:ρημ_τύπος
+    if t_name == "ρημ τύπος":
+        lemma = t_node.largs[-1][0]
+        assert isinstance(lemma, str)
```
Instead of asserting here, using `clean_node()` (no, wait, was it clean_node()?) to parse the contents of the argument is safe. Unlikely to matter here, but sometimes templates allow all sort of stuff in their arguments.

daxida replied:
```py
        lemma = t_node.largs[-1][0]
        lemma_bis = clean_node(wxr, None, t_node)
```

They seem to be different

```
 lemma='κόρφος' lemma_bis=':Πρότυπο:πτώσειςΟΚπλ'
 lemma='ξεκινώ' lemma_bis=':Πρότυπο:ρημ τύπος'
```

The first one just takes the last argument on the right of the templates (they all seem to be structured this way)

```
{{πτώσηΑεν|κόρφος}} > κόρφος
# {{ρημ τύπος|γ' ενικό [[οριστική]]ς αορίστου|ξεκινώ}} > ξεκινώ
# {{ρημ τύπος|β' ενικό [[προστακτική]]ς αορίστου|ξεκινώ}} > ξεκινώ
```

---

Not even this:

```py
        lemma = t_node.largs[-1][0]
        lemma_bis = clean_node(wxr, None, t_node.largs)
```
```
lemma='ξεκινώ' lemma_bis="ρημ τύποςα' ενικό οριστικής αορίστουξεκινώ"
```

kristian-clausal replied:
`lemma_bis = clean_node(wxr, None, tnode.largs[-1])` takes the argument (or should) and expands and parses and cleans it like wikitext, which gives you safe output that doesn't need to be asserted.

kristian-clausal commented:
Ok, this actually looked good. The return value stuff isn't vital here, looking at other sense entries we still have to parse for glosses so just inserting the `FormOf` data straight into the sense should be fine.

kristian-clausal commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -498,6 +498,86 @@ def bold_node_fn(
 
+def parse_gloss_inflections(
+    parent_sense: Sense, contents: list[str | WikiNode]
+) -> None:
```
Actually, this is fine. Looking at other sense entries with `form_of`, they do have glosses, so just letting the rest of the glosses stuff continue is fine and appropriate.

daxida commented:
The problem with tags instead of raw_tags is that they are deleted with the current code. One would need to change this

```py
def convert_tags_in_sense(sense: Sense) -> None:
    tags, raw_tags, poses = convert_tags(sense.raw_tags)
    # sense.tags = tags
    sense.tags.extend(tags) # <-
    sense.raw_tags = raw_tags
    sense.tags.extend(poses)
```

and I am not entirely sure of the side effects.

daxida commented on tests/test_el_inflection.py:
```diff
@@ -0,0 +1,119 @@
+from typing import Any
```
Same as above, I'm not entirely sure since it also adds (and tests for) tags. Other editions use this `test_X_inflections.py` name to.

But I am not specially attached to either name, if you really want me I could change both.

daxida commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -508,6 +588,8 @@ def parse_gloss(
     if len(contents) == 0:
         return False
 
+    parse_gloss_inflections(parent_sense, contents)
```
This was done in an earlier version were that function returned a bool if it parsed something and then it became:

```py
     if parse_gloss_inflections(parent_sense, contents):
          return True
```

but this deleted the gloss (since we exited the function because parsing it), and I discarded it. I saw no use of the return value since the rest of the `parse_gloss` function should still run no matter what.

daxida commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -498,6 +498,86 @@ def bold_node_fn(
+        # Cf. {{ουδ του-πτώσειςΟΑΚεν|καλός|grc}}
+        gender, inflection = t_name.split("-")
+        code = gender[:3]
+        GENDER_INFLECTION_MAP = {
```
I could do a match, but this small dict seemed to me like something you would like to have somewhere and reuse for different templates, instead of having a match statement at every function location where you find yourself facing the same problem.

Every CAPITALIZED_CONSTANT should probably be outside of the function at some point.

daxida commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -498,6 +498,86 @@ def bold_node_fn(
+    tags.extend([elt for elt in cases + [number]])
+    tags.sort()  # For the tests, but also good practice
+
+    lemma = only_node.largs[-1][0]
```
I commented below, but this seems to be a constant pattern of these templates. The lemma always come last (and for some reason they return it in a list? could they be more than one lemma? I don't think so but remains to be tested)

kristian-clausal replied:
The lemma may come last, but that might just means people write key-word arguments before the lemma, but `largs` is the actual raw arguments, which means they `key=arg` items might come after the lemma if someone does it that way. We should probably have come central `.get_args()` method for `TemplateNode`, now that we actually have a `TemplateNode` class... Especially now that other people are using the code.

daxida commented:
Addressed most of your suggestions. Let me know if you still want me to change the names of the functions/tests.

I will put this as ready because I won't be implementing more logic on my side other than the suggestions that you may still have.

daxida commented:
Note that I'm not entirely convinced about having to translate tags to English since this is supposed to be a monolingual el-el dictionary but oh well.

kristian-clausal commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -559,7 +559,11 @@ def parse_gloss_inflections_ptosi(
+        try:
+            gender_tag = GENDER_INFLECTION_MAP[code]
+        except KeyError:
+            # Bad template name. Can this even happen?
```
This does take in arbitrary template names with no guarantee for their structure. You'd need to be sure that `code = gender[:3]`is a key in `GENDER_INFLECTION_MAP`, and working with Wiktionary extraction before, I can guarantee you someone somewhere has made a template with a name that will break this.

kristian-clausal commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -498,6 +498,86 @@ def bold_node_fn(
+            "ουδ": "ουδέτερο",
+        }
+        gender_tag = GENDER_INFLECTION_MAP[code]
+        tags.append(gender_tag)
```
If you add data to `tags` it should be English tag strings, not Greek. You need to translate GENDER_INFLECTION_MAP values to their English tag counterparts: `"θηλ": "feminine",`

kristian-clausal replied:
You've changed it elsewhere, so this was just forgotten by mistake.

kristian-clausal commented on src/wiktextract/extractor/el/tags_utils.py:
```diff
@@ -65,6 +65,6 @@ def convert_tags(raw_tags: list[str]) -> tuple[list[str], list[str], list[str]]:
 def convert_tags_in_sense(sense: Sense) -> None:
     tags, raw_tags, poses = convert_tags(sense.raw_tags)
-    sense.tags = tags
+    sense.tags.extend(tags)
```
This is correct, the original was fine earlier, but it really should have been `.extends()` all along.

kristian-clausal commented:
> Note that I'm not entirely convinced about having to translate tags to English since this is supposed to be a monolingual el-el dictionary but oh well.

The Wiktextract project is for processing different Wiktionary editions and producing output with equivalent tags. This means translating tags into English.

kristian-clausal commented on src/wiktextract/extractor/el/pos.py:
```diff
@@ -498,6 +498,86 @@ def bold_node_fn(
     return None
 
 
+def parse_gloss_inflections(
```
`form_of` can also refer to nouns.

daxida replied:
Changed to `extract_form_of_templates` and `extract_form_of_templates_ptosi`

kristian-clausal commented:
I'll take a look at tomorrow. :+1:

kristian-clausal commented:
This works well enough for now, and I think we can merge it and handle any issue that come up later.

Something more useful than raw .largs should be implemented, for example.

kristian-clausal commented:
Wait, what the hell, there was already `template_parameters()`... Why haven't I noticed this before??? Why have I been messing around with `.largs` all this time? I must have forgotten this exists, when xxyzz implemented it!

daxida commented:
Ah, I didn't know about that. I saw your PR and I can understand the problem it solves. I will try to remember it for the future.

Thank you for the reviews and feedback, and have a nice vacation!

xxyzz commented:
> there was already `template_parameters()`... Why haven't I noticed this before??? 

Err, you made some changes to that class method [two moths ago](https://github.com/tatuylonen/wikitextprocessor/pull/363)... I guess you still influenced by en edition code?

kristian-clausal commented:
> > there was already `template_parameters()`... Why haven't I noticed this before???
> 
> Err, you made some changes to that class method [two months ago](https://github.com/tatuylonen/wikitextprocessor/pull/363)... I guess you still influenced by en edition code?

I already said I had forgotten about it, you don't have to mention it again.

xxyzz commented:
Sorry, just feel a bit surprising. I think I also forget many things...

kristian-clausal commented:
Sorry about being snippy, I [got up on the wrong side of the bed](https://en.wiktionary.org/wiki/get_up_on_the_wrong_side_of_the_bed).
